File: lib/token_lens/renderer/annotator.rb
# frozen_string_literal: true

module TokenLens
  module Renderer
    class Annotator
      def annotate(nodes, depth = 0)
        nodes.each do |node|
          node[:depth] = depth
          annotate(node[:children], depth + 1)
          child_tokens = node[:children].sum { |c| c[:subtree_tokens] }
          node[:subtree_tokens] = [node[:token].total_tokens, 1].max + child_tokens
        end
        nodes
      end
    end
  end
end

File: src/renderer/annotator.ts
import { costUsd, displayWidth } from "../token";
import type { Node } from "../types";

export function annotate(nodes: Node[], depth = 0): void {
  for (const node of nodes) {
    node.depth = depth;
    annotate(node.children, depth + 1);
    const childTokens = node.children.reduce((sum, c) => sum + (c.subtreeTokens ?? 0), 0);
    const childCost = node.children.reduce((sum, c) => sum + (c.subtreeCost ?? 0), 0);
    node.subtreeTokens = Math.max(displayWidth(node.token), 1) + childTokens;
    node.subtreeCost = costUsd(node.token) + childCost;
  }
}
